# Groupby

In [1]:
# importamos las librerías que necesitamos

# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np


# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

In [2]:
# cargamos el dataframe creado en la lección anterior
df = pd.read_csv("bank-additional_full.csv", index_col = 0)
df.head(1)

kidhome  teenhome dt_customer  numwebvisitsmonth  \
income                                                       
161770.0      1.0       0.0  2012-04-04               29.0   

                                            id  age        job  marital  \
income                                                                    
161770.0  089b39d8-e4d0-461b-87d4-814d71e0e079  NaN  housemaid  married   

         education  default  housing  loan    contact  duration  campaign  \
income                                                                      
161770.0  basic 4y      0.0      0.0   0.0  telephone       261         1   

          pdays  previous     poutcome  empvarrate conspriceidx consconfidx  \
income                                                                        
161770.0    999         0  nonexistent         1.1       93,994       -36,4   

         euribor3m nremployed   y           date  latitude  longitude  \
income                                                                  
161770.0     4,857       5191  no  2-agosto-2019    41.495    -71.233   

         contact_month  contact_year  
income                                
161770.0        agosto        2019.0

In [3]:
# contestemos a la primera pregunta, ¿cuál es la edad media de los clientes en función de su educación?
# por lo tanto, en este caso tendremos que agrupar por el tipo de educación y calcular la media de la edad usando el método '.mean()' 
# pero antes de calcular la media veamos que pasa si solo le pasamos el groupby sin especificar que operación queremos hacer (en nuestro caso la media)
df.groupby("education")

In [4]:
# en la línea de código anterior vemos que en realidad nos devuelve un objeto de tipo groupby, pero no nos da ningún resultado como tal. 
# y es que cuando hacemos un groupby y no le especificamos lo que queremos hacer/calcular, nos devolverá este objeto. 
# por lo tanto, añadamos a continuación lo que queremos calcular, en este caso la media
df.groupby("education").mean(numeric_only = True)

income   kidhome  teenhome  numwebvisitsmonth  \
education                                                                  
basic 4y             92337.678375  0.990817  1.000230          16.729109   
basic 6y             92193.844929  1.033110  0.989941          16.545683   
basic 9y             93563.364400  1.002853  0.998098          16.537328   
high school          94141.961411  1.007053  0.989521          16.496222   
illiterate           80480.388889  0.888889  0.888889          16.222222   
professional course  93669.489867  1.019536  1.005477          16.653825   
university degree    92738.460148  0.998349  1.002673          16.610753   

                           age   default   housing      loan    duration  \
education                                                                  
basic 4y             47.604900  0.000000  0.520425  0.152538  264.402204   
basic 6y             40.418783  0.000000  0.520448  0.145932  263.083822   
basic 9y             39.101327  0.000000  0.532423  0.149033  260.404977   
high school          37.909434  0.000121  0.528666  0.155702  260.479295   
illiterate           48.785714  0.000000  0.555556  0.166667  276.777778   
professional course  39.994828  0.000433  0.554039  0.158003  252.658755   
university degree    38.844152  0.000000  0.546566  0.162619  252.436567   

                     campaign       pdays  previous  empvarrate   latitude  \
education                                                                    
basic 4y             2.602388  965.710973  0.145317    0.211088  36.847573   
basic 6y             2.559514  980.689438  0.128667    0.244635  36.760509   
basic 9y             2.536852  979.181170  0.141702    0.149834  36.841785   
high school          2.562217  964.589723  0.187909    0.029602  36.884489   
illiterate           2.277778  943.833333  0.111111   -0.133333  35.827667   
professional course  2.594303  960.771043  0.164506    0.167866  36.794159   
university degree    2.559896  950.978620  0.194152   -0.034342  36.847812   

                      longitude  contact_year  
education                                      
basic 4y             -95.818800   2017.018933  
basic 6y             -95.722083   2016.950295  
basic 9y             -96.063955   2016.960993  
high school          -96.114177   2017.017828  
illiterate          -106.574167   2016.833333  
professional course  -95.933331   2017.010846  
university degree    -95.813060   2017.005695

In [5]:
# vemos que en el caso anterior, nos ha devuelto la media de todas las columnas numéricas. Pero a nosotras no nos interesa la media de todas las columnas, 
# SOLO NOS INTERESA LA MEDIA DE LA EDAD
# para eso tendremos que seguir la siguiente sintaxis, fijaos como le hemos pasado una lista fuera de groupby especificando la columna sobre la que quiero calcular la media
# en este caso vamos a almacenar los resultados del groupby en una variable
educacion_edad = df.groupby("education")["age"].mean()
educacion_edad

education
basic 4y               47.604900
basic 6y               40.418783
basic 9y               39.101327
high school            37.909434
illiterate             48.785714
professional course    39.994828
university degree      38.844152
Name: age, dtype: float64

In [6]:
# si nos fijamos, el resultado que nos devuelve es una Serie, pero a nosotros nos puede interesar tenerlo como un DataFrame. 
# para eso lo que tendremos que hacer es usar el método '.reset_index()' de Pandas y almacenarlo en una variable
df_educacion_edad = educacion_edad.reset_index()
df_educacion_edad

education        age
0             basic 4y  47.604900
1             basic 6y  40.418783
2             basic 9y  39.101327
3          high school  37.909434
4           illiterate  48.785714
5  professional course  39.994828
6    university degree  38.844152

In [7]:
# vale... ya hemos contestado a la primera pregunta, vamos a por la segunda:  ¿son los clientes que aceptan el producto ofrecido en la campaña de marketing más jovenes?
# para contestar a este pregunta tendré que agrupar por la columna "y" donde tenemos la información de si el cliente acepto el producto ofrecido en la campaña o no
# además tendremos que calcular la media de la edad, igual que en el ejemplo anterior. 
# recordad que añadimos el reset_index() al final para conseguir un DataFrame

# en este caso, podemos ver como no hay una gran diferencia entre las medias de las edades de nuestros clientes que han aceptado o no. 
df_edad_aceptacion = df.groupby("y")["age"].mean().reset_index()
df_edad_aceptacion

y        age
0   no  39.861724
1  yes  40.886310

In [8]:
# vamos con la tercera de las preguntas: ¿hay diferencia entre la cantidad de gente que acepta nuestra propuesta o no y su estado civil?
# en este caso están implicadas dos variables categóricas, lo que podría plantearnos si se puede agrupar por más de una columna, la respuesta es si.
# en este caso la sintaxis cambiará un poco y lo que tendremos que hacer es pasar una lista dentro del groupby con las columnas por las que queremos agrupar

# además ahora no queremos hacer una media si no que queremos contar el número de clientes que tenemos para cada combinación de categorías, 
# para eso lo que haremos será contar el número de id que tenemos
df.groupby(["marital", "y"])["id"].count().reset_index()

marital    y     id
0  divorced   no   4321
1  divorced  yes    490
2   married   no  23344
3   married  yes   2655
4    single   no  10419
5    single  yes   1686

Otros estadísticos que podemos calcular con el groupby son: 

- `count()` – número de observaciones no nulas
- `describe()` - resumen de los principales estadísticos
- `sum()` – suma de todos los valores
- `mean()` – media de los valores
- `median()` – mediana de los valores
- `min()` – valor mínimo
- `max()` – valor máximo
- `std()` – desviación estándar
- `var()`– varianza

Ampliemos un poco una de las preguntas planteadas al inicio del jupyter, "¿Cuál es la edad media de los clientes en función de su educación?". En este caso, no calcularemos solo la media, sino que calcularemos también la mediana, la varianza, la desviación estándar, el máximo y el mínimo. 


In [9]:
educacion_edad_ampliada = df.groupby("education")["age"].agg(["mean", "median", "std", "var", "min", "max"])
educacion_edad_ampliada

mean  median        std         var   min   max
education                                                                
basic 4y             47.604900    47.0  12.261436  150.342812  18.0  98.0
basic 6y             40.418783    39.0   8.680987   75.359532  18.0  95.0
basic 9y             39.101327    38.0   9.604425   92.244978  17.0  94.0
high school          37.909434    36.0   9.681421   93.729917  18.0  88.0
illiterate           48.785714    45.0  11.709159  137.104396  34.0  80.0
professional course  39.994828    38.0   9.869948   97.415864  20.0  86.0
university degree    38.844152    36.0   9.626975   92.678640  20.0  91.0

El método .agg() en Pandas se utiliza para aplicar múltiples funciones de agregación a los datos de un DataFrame. En el código que tenemos aqui arriba, se está utilizando agg para calcular varias estadísticas descriptivas sobre la columna age del DataFrame, en este caso la media, la mediana, la desviacion standard, la varianza, el minimo y el maximo.

Interpretemos los resultados obtenidos del gropuby anterior, pero antes de lanzarnos a ver los números entendamos que es esto de la media, mediana, desviación estándar(std) y varianza (var). 


1. Media: La media es el promedio de un conjunto de datos. Se obtiene sumando todos los valores y luego dividiendo esa suma entre el número total de datos.


    Interpretación: Supongamos que la media de edad de las alumnas es de aproximadamente 26.4 años. Es una medida central que nos da una idea general del valor típico de edad en el grupo.

2. Desviación estándar: La desviación estándar es una medida de dispersión que nos indica cuánto se desvían los valores individuales del promedio (la media). Una desviación estándar alta indica que los valores están más dispersos, mientras que una desviación estándar baja indica que los valores están más cercanos a la media.

    Interpretación: Supongamos que la desviación estándar resulta ser 2.5. Esto significa que, en promedio, las edades de las alumnas están dispersas alrededor de 2.5 años de distancia de la media (26.4). Siendo una desviación estándar relativamente baja, podemos decir que las edades de las alumnas están relativamente cercanas al valor promedio.

3. Varianza: La varianza es otra medida de dispersión que mide el grado en que los datos se alejan de la media. Se calcula como el promedio de los cuadrados de las diferencias entre cada valor y la media.


    Interpretación: Si suponemos que la varianza resulta ser 6.25, esto significa que, en promedio, las edades de las alumnas difieren aproximadamente 6.25 años cuadrados de la media (26.4). La varianza es útil para comprender la dispersión de los datos, pero puede ser difícil de interpretar directamente debido a que se expresan en unidades al cuadrado.

4. Mediana: La mediana es el valor que se encuentra justo en el centro de un conjunto de datos ordenados. Divide al conjunto en dos partes iguales, donde la mitad de los datos están por encima y la otra mitad por debajo.

    Interpretación: Supongamos que la mediana tiene un valor de 82. Esto nos indica que la mitad de las alumnas obtuvo una nota por encima de 82 y la otra mitad por debajo de este valor. Es útil para describir un valor típico en datos que pueden estar influenciados por valores extremos (outliers).

Estas medidas estadísticas son esenciales para resumir, describir y entender los datos en el contexto de la exploración y análisis de datos. Proporcionan una visión general de la distribución, la tendencia central y la dispersión de los datos, lo que permite tomar decisiones basadas en datos sólidos y obtener información valiosa para la toma de decisiones en diversas áreas como investigación, planificación empresarial, marketing, entre otros.

**¿Qué conclusiones podemos sacar de nuestro groupby?**

- La mediana vemos que su valor es de 47 años para los clientes que tienen una educación *basic 4y*, esto significa que el 50 % de los clientes con este tipo de educación tienen una edad menor a 47 años y el otro 50% tienen una edad mayor que 47.  

- Por otro lado vemos que de media los clientes con una educación de *high school* son los más jovenes (media de 37.90). Por el contrario, clientes con una educación de *illiterate* y *basic 4y* tienen unas medias muy parecidas siendo los que tienen una media mayor de edad. 

- La varianza nos indica la dispersión de los datos en unidades al cuadrado, cosa que se puede hacer bastante complejo de interpretar, por eso tenemos la desviación que ya no esta expresada en unidades al cuadrado y es más fácil. Vamos a centrarnos en una sola de las categorías, en concreto en *basic 4y* que tiene una media de 47.60 y una desviación de 12.26, esto significa que cada cliente se desvía de media 12.26 años respecto al valor de la media (12.26). 

- Ahora nos puede surgir la duda de **¿cómo podemos saber si esa desviación es grande o pequeña?** Para debemos compararla con la media y tener en cuenta el contexto específico del problema. Aquí hay algunas pautas generales para interpretar la magnitud de la desviación estándar:

    - Comparación con la media:

        - Si la desviación estándar es mucho menor que la media, indica que los datos tienden a estar muy cerca del valor promedio. En otras palabras, los datos están más agrupados y tienen poca dispersión alrededor de la media. Esto se consideraría una desviación estándar pequeña.

        - Si la desviación estándar es aproximadamente igual a la media, sugiere que los datos tienen una dispersión moderada alrededor del promedio. En este caso, se consideraría una desviación estándar de tamaño medio.

        - Si la desviación estándar es considerablemente mayor que la media, significa que los datos tienen una dispersión significativa alrededor del promedio. Esto se consideraría una desviación estándar grande.

    - Escala de los datos:

        - La magnitud absoluta de la desviación estándar depende de la escala de los datos. Si los valores en el conjunto de datos tienen una escala grande, la desviación estándar también será grande en términos absolutos. Del mismo modo, si los datos tienen una escala pequeña, la desviación estándar será pequeña en términos absolutos.

        - Por lo tanto, es importante considerar la magnitud de los datos y no solo el valor numérico de la desviación estándar.




## Otros métodos interesantes del `groupby()`

In [10]:
# recordamos el df
df.head(2)

income  kidhome  teenhome dt_customer  numwebvisitsmonth  \
0  161770        1         0  2012-04-04                 29   
1   85477        1         1  2012-12-30                  7   

                                     id   age        job  marital  \
0  089b39d8-e4d0-461b-87d4-814d71e0e079   NaN  housemaid  married   
1  e9d37224-cb6f-4942-98d7-46672963d097  57.0   services  married   

     education  default  housing  loan    contact  duration  campaign  pdays  \
0     basic 4y      0.0      0.0   0.0  telephone       261         1    999   
1  high school      NaN      0.0   0.0  telephone       149         1    999   

   previous     poutcome  empvarrate conspriceidx consconfidx euribor3m  \
0         0  nonexistent         1.1       93,994       -36,4     4,857   
1         0  nonexistent         1.1       93,994       -36,4       NaN   

  nremployed   y                date  latitude  longitude contact_month  \
0       5191  no       2-agosto-2019    41.495    -71.233        agosto   
1       5191  no  14-septiembre-2016    34.601    -83.923    septiembre   

   contact_year  
0        2019.0  
1        2016.0

In [11]:
# vamos a contar el número de personas que han aceptado o no el producto en función de la forma de contacto. 
# empezamos creando el groupby, y en este caso no le vamos a indicar que función de agregación queremos 
# YA QUE TODOS LOS MÉTODOS QUE VAMOS A APRENDER EN ESTE APARTADO SOLO FUNCIONAN SOBRE UN OBJETO DE TIPO GROUPBY
df_contact_acept = df.groupby(["contact", "y"])
df_contact_acept


In [12]:
# Método 'ngroups': Indica el número de grupos distintos creados después de agrupar los datos
# utilizando la cláusula groupby. 
# NOTA: este método se tiene que aplicar sobre un objeto de tipo groupby y NO sobre un DataFrame o una Serie
# en este caso nos dicen que se han generado 4 grupos. 
df_contact_acept.ngroups

4

In [13]:
# añadamos ahora la función de agregación para ver que grupos nos han quedado
# como nos piden saber el número de personas que han aceptado o no el producto en función de la forma de contacto 
# nuestra función de agregación será un '.count()'
df.groupby(["contact", "y"])['id'].count().reset_index()

contact    y     id
0   cellular   no  23357
1   cellular  yes   4039
2  telephone   no  14799
3  telephone  yes    805